deps/flatgeobuf/packedrtree.cpp

```diff
@@ -0,0 +1,410 @@
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+// NOTE: The upstream of this file is in https://github.com/bjornharrtell/flatgeobuf/tree/master/src/cpp
+
+#ifdef GDAL_COMPILATION
+#include "cpl_port.h"
+#else
+#define CPL_IS_LSB 1
+#endif
+
+#include "packedrtree.h"
+
+#include <map>
+#include <unordered_map>
+#include <iostream>
+
+namespace FlatGeobuf
+{
+
+const NodeItem &NodeItem::expand(const NodeItem &r)
+{
+    if (r.minX < minX) minX = r.minX;
+    if (r.minY < minY) minY = r.minY;
+    if (r.maxX > maxX) maxX = r.maxX;
+    if (r.maxY > maxY) maxY = r.maxY;
+    return *this;
+}
+
+NodeItem NodeItem::create(uint64_t offset)
+{
+    return {
+        std::numeric_limits<double>::infinity(),
+        std::numeric_limits<double>::infinity(),
+        -1 * std::numeric_limits<double>::infinity(),
+        -1 * std::numeric_limits<double>::infinity(),
+        offset
+    };
+}
+
+bool NodeItem::intersects(const NodeItem &r) const
+{
+    if (maxX < r.minX) return false;
+    if (maxY < r.minY) return false;
+    if (minX > r.maxX) return false;
+    if (minY > r.maxY) return false;
+    return true;
+}
+
+std::vector<double> NodeItem::toVector()
+{
+    return std::vector<double> { minX, minY, maxX, maxY };
+}
+
+// Based on public domain code at https://github.com/rawrunprotected/hilbert_curves
+uint32_t hilbert(uint32_t x, uint32_t y)
+{
+    uint32_t a = x ^ y;
+    uint32_t b = 0xFFFF ^ a;
+    uint32_t c = 0xFFFF ^ (x | y);
+    uint32_t d = x & (y ^ 0xFFFF);
+
+    uint32_t A = a | (b >> 1);
+    uint32_t B = (a >> 1) ^ a;
+    uint32_t C = ((c >> 1) ^ (b & (d >> 1))) ^ c;
+    uint32_t D = ((a & (c >> 1)) ^ (d >> 1)) ^ d;
+
+    a = A; b = B; c = C; d = D;
+    A = ((a & (a >> 2)) ^ (b & (b >> 2)));
+    B = ((a & (b >> 2)) ^ (b & ((a ^ b) >> 2)));
+    C ^= ((a & (c >> 2)) ^ (b & (d >> 2)));
+    D ^= ((b & (c >> 2)) ^ ((a ^ b) & (d >> 2)));
+
+    a = A; b = B; c = C; d = D;
+    A = ((a & (a >> 4)) ^ (b & (b >> 4)));
+    B = ((a & (b >> 4)) ^ (b & ((a ^ b) >> 4)));
+    C ^= ((a & (c >> 4)) ^ (b & (d >> 4)));
+    D ^= ((b & (c >> 4)) ^ ((a ^ b) & (d >> 4)));
+
+    a = A; b = B; c = C; d = D;
+    C ^= ((a & (c >> 8)) ^ (b & (d >> 8)));
+    D ^= ((b & (c >> 8)) ^ ((a ^ b) & (d >> 8)));
+
+    a = C ^ (C >> 1);
+    b = D ^ (D >> 1);
+
+    uint32_t i0 = x ^ y;
+    uint32_t i1 = b | (0xFFFF ^ (i0 | a));
+
+    i0 = (i0 | (i0 << 8)) & 0x00FF00FF;
+    i0 = (i0 | (i0 << 4)) & 0x0F0F0F0F;
+    i0 = (i0 | (i0 << 2)) & 0x33333333;
+    i0 = (i0 | (i0 << 1)) & 0x55555555;
+
+    i1 = (i1 | (i1 << 8)) & 0x00FF00FF;
+    i1 = (i1 | (i1 << 4)) & 0x0F0F0F0F;
+    i1 = (i1 | (i1 << 2)) & 0x33333333;
+    i1 = (i1 | (i1 << 1)) & 0x55555555;
+
+    uint32_t value = ((i1 << 1) | i0);
+
+    return value;
+}
+
+uint32_t hilbert(const NodeItem &r, uint32_t hilbertMax, const double minX, const double minY, const double width, const double height)
```

Comment:
we have it a bit differently in core postgis: https://github.com/postgis/postgis/blob/18ce69feba33b13d9fc6122dd594669d9b1bb977/liblwgeom/gbox.c#L796

Reply:
packedrtree.cpp is from upstream flatgeobuf and I would like to not have to customize that so I hope that duplication is ok.